README.md
# @valbo/auth-middlewares

Express middlewares to authenticate via basic auth or JWT and to authorize based on user role.

![npm (scoped)](https://img.shields.io/npm/v/@valbo/auth-middlewares)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
![Build Status](https://img.shields.io/github/workflow/status/valverdealbo/auth-middlewares/CI)
[![Coverage Status](https://coveralls.io/repos/github/valverdealbo/auth-middlewares/badge.svg?branch=main)](https://coveralls.io/github/valverdealbo/auth-middlewares?branch=main)
[![Known Vulnerabilities](https://snyk.io/test/github/valverdealbo/auth-middlewares/badge.svg?targetFile=package.json)](https://snyk.io/test/github/valverdealbo/auth-middlewares?targetFile=package.json)

## Contents

- [Install](#install)
- [Usage](#usage)
  - [Basic authentication](#basic-authentication)
  - [JWT authentication](#jwt-authentication)
  - [Role authorization](#role-authorization)
  - [Skipping middleware routes](#skipping-middleware-routes)
  - [Sending the user](#sending-the-user)
  - [Options](#options)

## Install

```bash
npm install @valbo/auth-middlewares
```

## Usage

This package exports express middlewares to authenticate via basic auth or JSON Web Token, and to authorize based on user roles. See the [example](src/example.ts) file for a full example on how to use all these middlewares. See the [options](#options) section at the end for an explanation on how to configure them.

### Basic authentication

There are two basic authentication middlewares: **askBasicAuth()** and **verifyBasicAuth()**.

**askBasicAuth()** checks for an **Authorization** header in the request and if it is missing it throws a **401** error and sets the **WWW-Authenticate** response header so that a browser asks for credentials. Always follow this middleware with **verifyBasicAuth()** so that when the user enters his credentials they will be verified.

**verifyBasicAuth()** validates an **Authorization** header with basic auth data. If the credentials are valid the authenticated user will be copied into **response.locals.user** and if the credentials are invalid it will throw a **401** error. The middleware will call **options.authenticationFunction** with the username and password it gets from the basic auth header.

### JWT authentication

There are two JWT authentication middlewares: **verifyTokenAuth()** and **setTokenAuth()**.

**verifyTokenAuth()** validates a JWT in a header. If the credentials are valid the authenticated user will be copied into **response.locals.user** and if the credentials are invalid it will throw a **401** error. The middleware will call **options.authenticationFunction** with a username (the **sub** claim of the JWT) and no password.

**setTokenAuth()** creates a JWT and sets it in a response header. The json web token **sub** claim is **response.locals.user.username**. It will throw a **500** error if there is any problem creating the token.

### Role authorization

**allowRoles()** checks if the authenticated user has one of the required roles. If it does not it will throw a **403** error. The middleware will call **options.authorizationFunction** with the user in **response.locals.user** and the requested roles.

### Skipping middleware routes

The **nextRouteIfUnauthorized()** and **nextRouteIfForbidden()** middlewares skip to the next middleware route if they receive a **401** and **403** errors respectively. They can be used to have both a private and a public version of the same route, or different versions of the same route for different roles.  

### Sending the user

**sendUser()** sends the user in **response.locals.user** as the response body or throws a **500** if there is no user in **response.locals**.

### Options

The **verifyBasicAuth()**, **verifyTokenAuth()**, **setTokenAuth()** and **allowRoles()** functions use the following options:

````typescript
export interface AuthenticationFunction {
  (username: string, password?: string): Promise<object | undefined>; // return the user if it is valid or undefined if not
}

export interface AuthorizationFunction<User extends object, Role extends string> {
  (user: User, ...roles: Role[]): boolean; // return whether the user has one of the requested roles
}

export interface BasicAuthenticationOptions {
  authenticationFunction: AuthenticationFunction; // will be called with both username and password
}

export interface TokenAuthenticationOptions {
  authenticationFunction: AuthenticationFunction; // will be called with only the username, since the JWT will already be valid
  jwtSigningSecret: string; // the secret to sign and verify JWT
  headerName: string; // the name of the header to use for the JWT, e.g. x-access-token
  expiresIn?: number; // if set the JWT will expire in X seconds
}

export interface AuthorizationOptions<User extends object, Role extends string> {
  authorizationFunction: AuthorizationFunction<User, Role>;
}

````
Each of the above functions actually receives either a regular object of the above types, or a function that receives the request and response and returns an options object:

```typescript
type Resolver<T> = T | ((request: Request, response: Response) => T);

function verifyBasicAuth(options: Resolver<BasicAuthenticationOptions>): RequestHandler;
function verifyTokenAuth(options: Resolver<TokenAuthenticationOptions>): RequestHandler;
function setTokenAuth(options: Resolver<TokenAuthenticationOptions>): RequestHandler;
function allowRoles<User extends object, Role extends string>(options: Resolver<AuthorizationOptions<User, Role>>,): (...roles: Role[]) => RequestHandler;
```

The functional form of the options can be used to dynamically decide which options to use, for example:

- Allowing only certain roles to log in depending on the origin of the request, like a website only for admin users.
- Using different JWT header names depending on the origin of the request, allowing users to log in simultaneously with multiple accounts.
- Returning a JWT with a different expiration depending on the role of the user.


CHANGELOG.md
## [1.0.5](https://github.com/valverdealbo/auth-middlewares/compare/v1.0.4...v1.0.5) (2021-04-18)


### Bug Fixes

* update dependencies ([154cf87](https://github.com/valverdealbo/auth-middlewares/commit/154cf879699a820d5d174a9f324dc65ee7cda311))

## [1.0.4](https://github.com/valverdealbo/auth-middlewares/compare/v1.0.3...v1.0.4) (2021-04-17)


### Bug Fixes

* update dependencies and tools configurations ([fb8c4b8](https://github.com/valverdealbo/auth-middlewares/commit/fb8c4b8a549dcc58635f86baf01c7a1eb1ee3c70))

## [1.0.3](https://github.com/valverdealbo/auth-middlewares/compare/v1.0.2...v1.0.3) (2021-03-23)


### Bug Fixes

* revert github action name ([331633d](https://github.com/valverdealbo/auth-middlewares/commit/331633d1f5a8efb649c20541876ce17a681bd579))

## [1.0.2](https://github.com/valverdealbo/auth-middlewares/compare/v1.0.1...v1.0.2) (2021-03-23)


### Bug Fixes

* update tooling configurations, update dependencies ([230b409](https://github.com/valverdealbo/auth-middlewares/commit/230b4092ccad31ef006a1b87ae03c8a8a1dcaa02))

## [1.0.1](https://github.com/valverdealbo/auth-middlewares/compare/v1.0.0...v1.0.1) (2021-02-18)


### Bug Fixes

* update dependencies ([00eefb1](https://github.com/valverdealbo/auth-middlewares/commit/00eefb1946b48a4cbaf2f6fbe5fdae61fba7c750))


src/index.test.ts
/* eslint-disable import/no-extraneous-dependencies,@typescript-eslint/no-explicit-any */
import express, { ErrorRequestHandler, RequestHandler } from 'express';
import supertest from 'supertest';
import * as jwt from 'jsonwebtoken';
import { InternalServerErrorError, UnauthorizedError, ForbiddenError } from '@valbo/http-errors';
import {
  AuthenticationFunction,
  AuthorizationFunction,
  askBasicAuth,
  verifyBasicAuth,
  verifyTokenAuth,
  setTokenAuth,
  allowRoles,
  nextRouteIfUnauthorized,
  nextRouteIfForbidden,
  sendUser,
} from '.';

const username = 'alice';
const password = '12345678';
const jwtSigningSecret = 'jwtSigningSecret';
const headerName = 'x-access-token';
const expiresIn = 24 * 60 * 60;

const authenticationFunctionUser: AuthenticationFunction = async () => ({ username, password });
const authenticationFunctionUndefined: AuthenticationFunction = async () => undefined;
const authenticationFunctionThrow: AuthenticationFunction = async () => {
  throw new InternalServerErrorError();
};

const setUser: (user: any) => RequestHandler = user => (request, response, next) => {
  response.locals.user = user;
  next();
};

const throwError: (error: any) => RequestHandler = error => () => {
  throw error;
};

type UserRole = 'admin' | 'dev' | 'user';

interface User {
  roles: UserRole[];
}

const authorizationFunction: AuthorizationFunction<User, UserRole> = (user, ...roles) => roles.some(role => user.roles.includes(role));

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const sendError: ErrorRequestHandler = (error, request, response, next) => {
  const httpError = {
    status: error.status || 500,
    name: error.name || InternalServerErrorError.name,
    message: error.message || '',
  };
  response.status(httpError.status).json({ error: httpError });
};

const sendEmpty: RequestHandler = (request, response) => {
  response.json({});
};

const authorizationBasicOk = 'Basic YWxpY2U6MTIzNDU2Nzg=';
const authorizationBasicError = '';
const tokenOk = jwt.sign({}, jwtSigningSecret, { subject: username });
const tokenError = jwt.sign({}, 'wrongSecret', { subject: username });
const tokenExpired = jwt.sign({ exp: Math.floor(Date.now() / 1000) - 60 }, jwtSigningSecret, { subject: username });

describe('askBasicAuth()', () => {
  test('should set the www-authenticate header and next an error when the authorization header is missing', async () => {
    const app = express().use(askBasicAuth()).use(sendEmpty).use(sendError);
    const response = await supertest(app).get('/');
    expect(response.header['www-authenticate']).toBeDefined();
    expect(response.status).toBe(401);
    expect(response.body.error.name).toBe(UnauthorizedError.name);
  });

  test('should next() when the authorization header exists', async () => {
    const app = express().use(askBasicAuth()).use(sendEmpty).use(sendError);
    const response = await supertest(app).get('/').set('Authorization', authorizationBasicOk);
    expect(response.status).toBe(200);
  });
});

describe('verifyBasicAuth()', () => {
  test('should next() an invalid credentials error when the authorization header is missing', async () => {
    const app = express()
      .use(verifyBasicAuth({ authenticationFunction: authenticationFunctionUser }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/');
    expect(response.status).toBe(401);
    expect(response.body.error.name).toBe(UnauthorizedError.name);
  });

  test('should next() an invalid credentials error when basicAuth cannot decode the authorization header', async () => {
    const app = express()
      .use(verifyBasicAuth({ authenticationFunction: authenticationFunctionUser }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set('Authorization', authorizationBasicError);
    expect(response.status).toBe(401);
    expect(response.body.error.name).toBe(UnauthorizedError.name);
  });

  test('should next() the error thrown by the authentication function', async () => {
    const app = express()
      .use(verifyBasicAuth({ authenticationFunction: authenticationFunctionThrow }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set('Authorization', authorizationBasicOk);
    expect(response.status).toBe(500);
    expect(response.body.error.name).toBe(InternalServerErrorError.name);
  });

  test('should next() an invalid credentials error when the authentication function returns undefined', async () => {
    const app = express()
      .use(verifyBasicAuth({ authenticationFunction: authenticationFunctionUndefined }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set('Authorization', authorizationBasicOk);
    expect(response.status).toBe(401);
    expect(response.body.error.name).toBe(UnauthorizedError.name);
  });

  test('should next() when the authentication function returns an object', async () => {
    const app = express()
      .use(verifyBasicAuth({ authenticationFunction: authenticationFunctionUser }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set('Authorization', authorizationBasicOk);
    expect(response.status).toBe(200);
    expect(response.body).toEqual({});
  });

  test('should work when called with an options resolver', async () => {
    const app = express()
      .use(verifyBasicAuth(() => ({ authenticationFunction: authenticationFunctionUser })))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set('Authorization', authorizationBasicOk);
    expect(response.status).toBe(200);
    expect(response.body).toEqual({});
  });
});

describe('verifyTokenAuth()', () => {
  const authOptions = { jwtSigningSecret, headerName };

  test('should next() a missing credentials error when the access header is missing', async () => {
    const app = express()
      .use(verifyTokenAuth({ authenticationFunction: authenticationFunctionUser, ...authOptions }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/');
    expect(response.status).toBe(401);
    expect(response.body.error.name).toBe(UnauthorizedError.name);
  });

  test('should next() an invalid credentials error when jwt.verify() throws a regular error', async () => {
    const app = express()
      .use(verifyTokenAuth({ authenticationFunction: authenticationFunctionUser, ...authOptions }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set(headerName, tokenError);
    expect(response.status).toBe(401);
    expect(response.body.error.name).toBe(UnauthorizedError.name);
  });

  test('should next() an expired credentials error when jwt.verify() throws an TokenExpiredError', async () => {
    const app = express()
      .use(verifyTokenAuth({ authenticationFunction: authenticationFunctionUser, ...authOptions }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set(headerName, tokenExpired);
    expect(response.status).toBe(401);
    expect(response.body.error.name).toBe(UnauthorizedError.name);
  });

  test('should next() the error thrown by the authentication function', async () => {
    const app = express()
      .use(verifyTokenAuth({ authenticationFunction: authenticationFunctionThrow, ...authOptions }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set(headerName, tokenOk);
    expect(response.status).toBe(500);
    expect(response.body.error.name).toBe(InternalServerErrorError.name);
  });

  test('should next() an invalid credentials error when the authentication function returns undefined', async () => {
    const app = express()
      .use(verifyTokenAuth({ authenticationFunction: authenticationFunctionUndefined, ...authOptions }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set(headerName, tokenOk);
    expect(response.status).toBe(401);
    expect(response.body.error.name).toBe(UnauthorizedError.name);
  });

  test('should next() when the authentication function returns an object', async () => {
    const app = express()
      .use(verifyTokenAuth({ authenticationFunction: authenticationFunctionUser, ...authOptions }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set(headerName, tokenOk);
    expect(response.status).toBe(200);
    expect(response.body).toEqual({});
  });

  test('should work when called with an options resolver', async () => {
    const app = express()
      .use(verifyTokenAuth(() => ({ authenticationFunction: authenticationFunctionUser, ...authOptions })))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set(headerName, tokenOk);
    expect(response.status).toBe(200);
    expect(response.body).toEqual({});
  });
});

describe('setTokenAuth()', () => {
  const authOptions = { jwtSigningSecret, headerName };

  test('should throw when response.locals.user.username is undefined', async () => {
    const app = express()
      .use(setTokenAuth({ authenticationFunction: authenticationFunctionUser, ...authOptions }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set(headerName, tokenOk);
    expect(response.status).toBe(500);
    expect(response.body.error.name).toBe(InternalServerErrorError.name);
  });

  test('should put the jwt in the header and call next()', async () => {
    const app = express()
      .use(verifyTokenAuth({ authenticationFunction: authenticationFunctionUser, ...authOptions }))
      .use(setTokenAuth({ authenticationFunction: authenticationFunctionUser, ...authOptions }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set(headerName, tokenOk);
    expect(response.status).toBe(200);
    const token = response.header[headerName];
    expect(token).toBeDefined();
    const payload = jwt.verify(token, jwtSigningSecret) as { sub: string };
    expect(payload.sub).toBe(username);
  });

  test('should set the exp claim in the token when expiresIn is provided', async () => {
    const app = express()
      .use(verifyTokenAuth({ authenticationFunction: authenticationFunctionUser, ...authOptions }))
      .use(setTokenAuth({ authenticationFunction: authenticationFunctionUser, ...authOptions, expiresIn }))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set(headerName, tokenOk);
    expect(response.status).toBe(200);
    const token = response.header[headerName];
    expect(token).toBeDefined();
    const payload = jwt.verify(token, jwtSigningSecret) as { exp: number; iat: number };
    expect(payload.exp - payload.iat).toBe(expiresIn);
  });

  test('should work when called with an options resolver', async () => {
    const app = express()
      .use(verifyTokenAuth({ authenticationFunction: authenticationFunctionUser, ...authOptions }))
      .use(setTokenAuth(() => ({ authenticationFunction: authenticationFunctionUser, ...authOptions })))
      .use(sendEmpty)
      .use(sendError);
    const response = await supertest(app).get('/').set(headerName, tokenOk);
    expect(response.status).toBe(200);
    const token = response.header[headerName];
    expect(token).toBeDefined();
    const payload = jwt.verify(token, jwtSigningSecret) as { sub: string };
    expect(payload.sub).toBe(username);
  });
});

describe('allowRoles()', () => {
  test('should throw an internal server error when response.locals.user is missing', async () => {
    const allow = allowRoles<User, UserRole>({ authorizationFunction });
    const app = express().use(allow('admin')).use(sendError).use(sendEmpty);
    const response = await supertest(app).get('/');
    expect(response.status).toBe(500);
  });

  test('should throw a forbidden error when the authorization function returns false', async () => {
    const allow = allowRoles<User, UserRole>({ authorizationFunction });
    const app = express()
      .use(setUser({ roles: [] }))
      .use(allow('admin'))
      .use(sendError)
      .use(sendEmpty);
    const response = await supertest(app).get('/');
    expect(response.status).toBe(403);
  });

  test('should next when the authorization function returns true', async () => {
    const allow = allowRoles<User, UserRole>({ authorizationFunction });
    const app = express()
      .use(setUser({ roles: ['admin'] }))
      .use(allow('admin'))
      .use(sendError)
      .use(sendEmpty);
    const response = await supertest(app).get('/');
    expect(response.status).toBe(200);
  });
});

describe('nextRouteIfUnauthorized()', () => {
  test('should capture an unauthorized error and skip to the next route', async () => {
    const app = express().use(throwError(new UnauthorizedError()), nextRouteIfUnauthorized()).use(sendError).use(sendEmpty);
    const response = await supertest(app).get('/');
    expect(response.status).toBe(200);
  });

  test('should send again any error other than unauthorized', async () => {
    const app = express().use(throwError(new InternalServerErrorError()), nextRouteIfUnauthorized(), sendError, sendEmpty);
    const response = await supertest(app).get('/');
    expect(response.status).toBe(500);
  });
});

describe('nextRouteIfForbidden()', () => {
  test('should capture a forbidden error and skip to the next route', async () => {
    const app = express().use(throwError(new ForbiddenError()), nextRouteIfForbidden()).use(sendError).use(sendEmpty);
    const response = await supertest(app).get('/');
    expect(response.status).toBe(200);
  });

  test('should send again any error other than forbidden', async () => {
    const app = express().use(throwError(new InternalServerErrorError()), nextRouteIfForbidden(), sendError, sendEmpty);
    const response = await supertest(app).get('/');
    expect(response.status).toBe(500);
  });
});

describe('sendUser()', () => {
  test('should send the user in the response', async () => {
    const app = express()
      .use((request, response, next) => {
        response.locals.user = { username };
        next();
      })
      .use(sendUser());
    const response = await supertest(app).get('/');
    expect(response.status).toBe(200);
    expect(response.body.user.username).toBe(username);
  });

  test('should throw an internal server error error when response.locals.user is missing', async () => {
    const app = express().use(sendUser()).use(sendError);
    const response = await supertest(app).get('/');
    expect(response.status).toBe(500);
    expect(response.body.error.name).toBe(InternalServerErrorError.name);
  });
});


src/example.ts
// eslint-disable-next-line import/no-extraneous-dependencies
import express, { Request, Response } from 'express';
import {
  askBasicAuth,
  verifyBasicAuth,
  verifyTokenAuth,
  setTokenAuth,
  allowRoles,
  nextRouteIfUnauthorized,
  nextRouteIfForbidden,
  sendUser,
  AuthenticationFunction,
  BasicAuthenticationOptions,
  TokenAuthenticationOptions,
  AuthorizationFunction,
  AuthorizationOptions,
} from '.';

type User = {
  username: string;
  password: string;
  role: 'admin' | 'user';
};

const users: User[] = [
  { username: 'admin', password: '1234', role: 'admin' },
  { username: 'user', password: '5678', role: 'user' },
];

const authenticateUser: AuthenticationFunction = async (username: string, password?: string) =>
  users.find(user => user.username === username && (password === undefined || user.password === password));

const authorizeUser: AuthorizationFunction<User, User['role']> = (user, ...roles) => roles.some(role => user.role === role);

const options: BasicAuthenticationOptions & TokenAuthenticationOptions & AuthorizationOptions<User, User['role']> = {
  authenticationFunction: authenticateUser,
  authorizationFunction: authorizeUser,
  jwtSigningSecret: '1d2c468e-3b6b-4309-9b4c-04a890da7e76',
  headerName: 'x-access-token',
  expiresIn: 24 * 60 * 60,
};

const auth = {
  askBasicAuth: askBasicAuth(),
  verifyBasicAuth: verifyBasicAuth(options),
  verifyTokenAuth: verifyTokenAuth(options),
  setTokenAuth: setTokenAuth(options),
  allowRoles: allowRoles<User, User['role']>(options),
  nextRouteIfUnauthorized: nextRouteIfUnauthorized(),
  nextRouteIfForbidden: nextRouteIfForbidden(),
  sendUser: sendUser(),
};

const app = express();

// login with basic auth, send back user and JWT in header for future requests
app.get('/login', auth.askBasicAuth, auth.verifyBasicAuth, auth.setTokenAuth, auth.sendUser);

// private route for admins
app.get(
  '/data',
  auth.verifyTokenAuth,
  auth.nextRouteIfUnauthorized,
  auth.allowRoles('admin'),
  auth.nextRouteIfForbidden,
  (request: Request, response: Response) => {
    response.json({ data: 'admin private data' });
  },
);

// private route for users
app.get(
  '/data',
  auth.verifyTokenAuth,
  auth.nextRouteIfUnauthorized,
  auth.allowRoles('user'),
  auth.nextRouteIfForbidden,
  (request: Request, response: Response) => {
    response.json({ data: 'user private data' });
  },
);

// public route
app.get('/data', (request: Request, response: Response) => {
  response.json({ data: 'public data' });
});


src/index.ts
/* eslint-disable @typescript-eslint/ban-types */
import { promisify } from 'util';
import type { ErrorRequestHandler, Request, Response, RequestHandler } from 'express';
import basicAuth from 'basic-auth';
import * as jwt from 'jsonwebtoken';
import { asyncMiddleware } from '@valbo/async-middleware';
import { InternalServerErrorError, UnauthorizedError, ForbiddenError } from '@valbo/http-errors';

const jwtSign = promisify<object, string, jwt.SignOptions, string>(jwt.sign);
const jwtVerify = promisify<string, string, jwt.VerifyOptions, { sub: string }>(jwt.verify);

export type Resolver<T> = T | ((request: Request, response: Response) => T);

function isResolved<T>(resolver: Resolver<T>): resolver is T {
  return typeof resolver !== 'function';
}

export function resolve<T>(request: Request, response: Response, resolver: Resolver<T>): T {
  return isResolved(resolver) ? resolver : resolver(request, response);
}

// will be called to authenticate a user, e.g. a database function
export interface AuthenticationFunction {
  (username: string, password?: string): Promise<object | undefined>;
}

// will be called to authorize a user, e.g. if the user has the required role
export interface AuthorizationFunction<User extends object, Role extends string> {
  (user: User, ...roles: Role[]): boolean;
}

export interface BasicAuthenticationOptions {
  authenticationFunction: AuthenticationFunction;
}

export interface TokenAuthenticationOptions {
  authenticationFunction: AuthenticationFunction;
  jwtSigningSecret: string;
  headerName: string; // the name of the header to use for authentication, e.g. x-access-token
  expiresIn?: number; // if set the token in the header will expire in X seconds
}

export interface AuthorizationOptions<User extends object, Role extends string> {
  authorizationFunction: AuthorizationFunction<User, Role>;
}

export function askBasicAuth(): RequestHandler {
  return asyncMiddleware(async (request, response, next) => {
    if (!('authorization' in request.headers)) {
      response.setHeader('WWW-Authenticate', `Basic realm="${request.hostname}"`);
      throw new UnauthorizedError('missing credentials in Authorization header');
    }
    next();
  });
}

export function verifyBasicAuth(options: Resolver<BasicAuthenticationOptions>): RequestHandler {
  return asyncMiddleware(async (request, response, next) => {
    const { authenticationFunction } = resolve(request, response, options);
    const nameAndPass = basicAuth(request);
    if (nameAndPass === undefined) {
      throw new UnauthorizedError('invalid Authorization basic auth header');
    }
    const user = await authenticationFunction(nameAndPass.name, nameAndPass.pass);
    if (user === undefined) {
      throw new UnauthorizedError('invalid username or password');
    }
    response.locals.user = user;
    next();
  });
}

export function verifyTokenAuth(options: Resolver<TokenAuthenticationOptions>): RequestHandler {
  return asyncMiddleware(async (request, response, next) => {
    const { authenticationFunction, jwtSigningSecret, headerName } = resolve(request, response, options);
    const header = request.headers[headerName] as string | undefined;
    if (header === undefined) {
      throw new UnauthorizedError(`missing ${headerName} header`);
    }
    let sub;
    try {
      ({ sub } = await jwtVerify(header, jwtSigningSecret, {}));
    } catch (error) {
      if (error.name === 'TokenExpiredError') {
        throw new UnauthorizedError(`expired token in ${headerName} header`);
      }
      throw new UnauthorizedError(`invalid token in ${headerName} header`);
    }
    const user = await authenticationFunction(sub);
    if (user === undefined) {
      throw new UnauthorizedError(`invalid username in ${headerName} header token`);
    }
    response.locals.user = user;
    next();
  });
}

export function setTokenAuth(options: Resolver<TokenAuthenticationOptions>): RequestHandler {
  return asyncMiddleware(async (request, response, next) => {
    const { jwtSigningSecret, headerName, expiresIn } = resolve(request, response, options);
    const username: string | undefined = response.locals.user?.username;
    if (username === undefined) {
      throw new InternalServerErrorError('missing response.locals.user.username');
    }
    const signOptions: jwt.SignOptions = { subject: username };
    if (expiresIn !== undefined) {
      signOptions.expiresIn = expiresIn;
    }
    const token = await jwtSign({}, jwtSigningSecret, signOptions);
    response.setHeader(headerName, token);
    next();
  });
}

export function allowRoles<User extends object, Role extends string>(
  options: Resolver<AuthorizationOptions<User, Role>>,
): (...roles: Role[]) => RequestHandler {
  return (...roles: Role[]): RequestHandler => (request, response, next): void => {
    const { authorizationFunction } = resolve(request, response, options);
    if (response.locals.user === undefined) {
      throw new InternalServerErrorError('missing response.locals.user');
    } else if (authorizationFunction(response.locals.user, ...roles)) {
      next();
    } else {
      next(new ForbiddenError(`user does not have the required role`));
    }
  };
}

export function nextRouteIfUnauthorized(): ErrorRequestHandler {
  return (error, request, response, next): void => {
    if (error.status === 401) {
      next('route');
    } else {
      next(error);
    }
  };
}

export function nextRouteIfForbidden(): ErrorRequestHandler {
  return (error, request, response, next): void => {
    if (error.status === 403) {
      next('route');
    } else {
      next(error);
    }
  };
}

export function sendUser(): RequestHandler {
  return asyncMiddleware(async (request, response) => {
    const user = response.locals.user as object | undefined;
    if (user === undefined) {
      throw new InternalServerErrorError('missing response.locals.user');
    }
    response.json({ user });
  });
}
